lib/src/executor/runtime_host.rs

```diff
@@ -1310,12 +1435,39 @@ impl Inner {
                     });
                 }
 
+                host::HostVm::ExternalOffchainStorageGet(req) => {
+                    let overlay = self.offchain_storage_changes.get(req.key().as_ref());
+                    match overlay {
+                        Some(value) => self.vm = req.resume(value.as_ref().map(|v| &v[..])),
+                        None => {
+                            self.vm = req.into();
+                            return RuntimeHostVm::Offchain(OffchainContext::StorageGet(
+                                OffchainStorageGet { inner: self },
+                            ));
+                        }
+                    }
+                }
+
                 host::HostVm::ExternalOffchainStorageSet(req) => {
-                    self.offchain_storage_changes.insert(
-                        req.key().as_ref().to_vec(),
-                        req.value().map(|v| v.as_ref().to_owned()),
-                    );
-                    self.vm = req.resume();
+                    let overlay = self.offchain_storage_changes.get(req.key().as_ref());
+
+                    let replace = match (overlay, req.old_value()) {
+                        (Some(Some(overlay)), Some(old_value)) => {
+                            old_value.as_ref().to_vec().eq(overlay)
+                        }
+                        _ => true,
+                    };
+
+                    if replace {
+                        if let Some(value) = req.value() {
+                            self.offchain_storage_changes
+                                .insert(req.key().as_ref().to_vec(), Some(value.as_ref().to_vec()));
+                        } else {
+                            self.offchain_storage_changes.remove(req.key().as_ref());
+                        }
+                    }
```

Comment:
That's not correct unfortunately for the compare-and-set variant.
If there's no existing value in the overlay, we have to query the current value from the API user in order to compare it with what `old_value` returns.

Please either implement properly or remove support for the `compare_and_set` function. :pray:

Reply:
oh that's right. We need `compare_and_set`, so all I need is insert value to `offchain_storage_changes` when it's injected to `OffchainStorageGet`

Reply:
Maybe I'm misunderstanding, but to me `compare_and_set` means: load the current value (with an `OffchainStorageGet` if necessary), compare with the value that the runtime provides, and only if they're different then write the new value that the runtime provides.

Reply:
Actually naming `overlay` is not correct (updating) because offchain worker doesn't have overlay, just read/write from storage.
The `old_value` provided by runtime is read using `OffchainStorageGet` first which is inserted into `offchain_storage_chages` https://github.com/smol-dot/smoldot/pull/788/files#diff-3cdf7d07bd6351e0742505e7dc4e11464f5c12424357a082ae6b3249d8b28298R932-R935

Reply:
Sorry for the delay.

After looking into Substrate, the `ext_offchain_index` functions are affected by the storage transactions: https://github.com/paritytech/substrate/blob/689da495a0c0c0c2466fe90a9ea187ce56760f2d/primitives/state-machine/src/overlayed_changes/mod.rs#L790-L843

However, the `ext_offchain_local_storage` functions indeed don't have an overlay and just directly load and dump from the database.

Reply:
I didn't check `ext_offchain_index`, I thought it was the same as `local_storage`. I will have another look

Reply:
ok so `ext_offchain_index_set/clear` is a way for on-chain to modify off-chain storage if transaction is successful and that's why this writes to overlay because it is committed or reverted the same as transaction
https://github.com/paritytech/substrate/blob/7b1d6f337fd1cee4e1fe675c89c8d82ce56d5c47/frame/session/src/historical/offchain.rs#L215

Reply:
@tomaka I suggest we update `ext_offchain_index_set/clear` methods to skip and continue execution and address this later.

Reply:
I don't think it's very complicated to implement properly:

- In `host.rs`, add back a variant for `ext_offchain_index_set/clear`, not the same as the one for `ext_offchain_local_storage_*` so that `runtime_host.rs` can know the difference.
- In `runtime_host.rs`, remove `offchain_storage_changes` from `Config` and `Success`.
- Add a `OffchainStorageSet` variant in `runtime_host`.
- Add a field `offchain_storage_changes` to `PendingStorageChanges` (`PendingStorageChanges` stores the state during a transaction).
- Around [here](https://github.com/smol-dot/smoldot/blob/7f6a4662a25b482f2fe8b991c291a39bfe31dbef/lib/src/executor/runtime_host.rs#L1088), add a `if matches!(self.vm, HostVm::Finished) && !self.offchain_storage_changes.is_empty() { ... }`, and in the `if` you remove an element from `offchain_storage_changes` and yield a `OffchainStorageSet`. This commits the overlay offchain storage changes.

Reply:
I added `offchain_index` into `PendingStorageChanges` and they're applied into `offchain_storage_changes` on success.